File: common/src/main/java/com/obsidian/backup/common/IpcProtocol.java
package com.obsidian.backup.common;

import com.google.gson.*;
import java.util.*;

/**
 * Defines the IPC message protocol between the Minecraft mod/plugin (Java)
 * and the Obsidian Sidecar daemon (Rust) over Unix Domain Socket.
 *
 * This class is loader-agnostic and version-independent — it depends only
 * on Gson and the Java standard library (Java 16+ for UnixDomainSocket).
 */
public final class IpcProtocol {

    public static final Gson GSON = new GsonBuilder().setPrettyPrinting().create();

    /** IPC operation codes matching the Sidecar's op dispatch table. */
    public enum OpCode {
        BACKUP("backup"),
        STATUS("status"),
        RESTORE("restore"),
        DIFF("diff"),
        BROWSE("browse"),
        TOP("top"),
        VERIFY("verify"),
        PIN("pin"),
        CLONE("clone"),
        ROLLBACK("rollback"),
        CANCEL("cancel"),
        FORECAST("forecast"),
        EXPORT("export"),
        IMPORT("import"),
        AUTH("auth");

        public final String code;

        OpCode(String code) {
            this.code = code;
        }
    }

    /** IPC request sent from mod → sidecar. */
    public static class Request {
        public String tx_id;
        public String op;
        public JsonObject params;

        public Request() {}

        public Request(String op, Map<String, Object> params) {
            this.tx_id = UUID.randomUUID().toString().substring(0, 8);
            this.op = op;
            this.params = paramsToJson(params);
        }

        private static JsonObject paramsToJson(Map<String, Object> params) {
            JsonObject obj = new JsonObject();
            if (params != null) {
                for (var entry : params.entrySet()) {
                    Object v = entry.getValue();
                    if (v instanceof String) obj.addProperty(entry.getKey(), (String) v);
                    else if (v instanceof Number) obj.addProperty(entry.getKey(), (Number) v);
                    else if (v instanceof Boolean) obj.addProperty(entry.getKey(), (Boolean) v);
                    else if (v instanceof JsonElement) obj.add(entry.getKey(), (JsonElement) v);
                }
            }
            return obj;
        }
    }

    /** IPC response sent from sidecar → mod. */
    public static class Response {
        public String tx_id;
        public String status;    // "ok" | "error" | "progress"
        public String message;
        public JsonObject data;
    }

    /** Status data from the Sidecar. */
    public static class StatusData {
        public boolean running;
        public String current_tx;
        public String state;
        public double tps;
        public double cpu_percent;
        public long memory_mb;
        public long disk_iops_read;
        public long disk_iops_write;
        public double network_upload_mbps;
        public QueueStatus queue_status;
        public StorageStats storage_stats;

        public static StatusData fromJson(JsonObject data) {
            return GSON.fromJson(data, StatusData.class);
        }
    }

    public static class QueueStatus {
        public int scanner;
        public int chunk;
        public int compress;
        public int encrypt;
        public int upload;
    }

    public static class StorageStats {
        public long total_snapshots;
        public long total_size_bytes;
        public double dedup_ratio;
        public long packfile_count;
    }

    // --- Serialization ---

    public static String toJson(Object obj) {
        return GSON.toJson(obj);
    }

    public static Response parseResponse(String json) {
        return GSON.fromJson(json, Response.class);
    }

    // --- Parameter builders ---

    public static Map<String, Object> paramsBackup(String tag, boolean incremental) {
        var params = new LinkedHashMap<String, Object>();
        params.put("tag", tag);
        params.put("incremental", incremental);
        params.put("timestamp", System.currentTimeMillis());
        return params;
    }

    public static Map<String, Object> paramsStatus() {
        return Collections.emptyMap();
    }

    public static Map<String, Object> paramsRestore(String snapshotId, String filePath, String chunkCoord) {
        var params = new LinkedHashMap<String, Object>();
        params.put("snapshot_id", snapshotId);
        if (filePath != null) params.put("file_path", filePath);
        if (chunkCoord != null) params.put("chunk_coord", chunkCoord);
        return params;
    }

    public static Map<String, Object> paramsTop(int limit) {
        return Map.of("limit", limit);
    }

    public static Map<String, Object> paramsDiff(String idA, String idB) {
        return Map.of("id_a", idA, "id_b", idB);
    }

    public static Map<String, Object> paramsBrowse(String snapshotId, String path) {
        var params = new LinkedHashMap<String, Object>();
        params.put("snapshot_id", snapshotId);
        if (path != null) params.put("path", path);
        return params;
    }

    public static Map<String, Object> paramsVerify(boolean repair) {
        return Map.of("repair", repair);
    }

    public static Map<String, Object> paramsPin(String snapshotId, int days) {
        return Map.of("snapshot_id", snapshotId, "days", days);
    }

    public static Map<String, Object> paramsClone(String snapshotId, String newName) {
        return Map.of("snapshot_id", snapshotId, "new_name", newName);
    }

    public static Map<String, Object> paramsRollback(String duration) {
        return Map.of("duration", duration);
    }

    public static Map<String, Object> paramsForecast() {
        return Collections.emptyMap();
    }

    public static Map<String, Object> paramsExport(String path) {
        return Map.of("path", path);
    }

    public static Map<String, Object> paramsImport(String path) {
        return Map.of("path", path);
    }

    public static Map<String, Object> paramsAuth(String token) {
        return Map.of("token", token);
    }
}

File: mod-neoforge/src/main/kotlin/com/obsidian/backup/ipc/IpcProtocol.kt
package com.obsidian.backup.ipc

import com.google.gson.Gson
import com.google.gson.GsonBuilder
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import com.google.gson.annotations.SerializedName
import java.util.UUID

/**
 * Defines the IPC message protocol between the Minecraft mod (Java/Kotlin)
 * and the Obsidian Sidecar daemon (Rust) over Unix Domain Socket.
 */
object IpcProtocol {

    val GSON: Gson = GsonBuilder()
        .setPrettyPrinting()
        .create()

    // --- Operation Codes ---
    enum class OpCode(val code: String) {
        @SerializedName("backup")  BACKUP("backup"),
        @SerializedName("status")  STATUS("status"),
        @SerializedName("restore") RESTORE("restore"),
        @SerializedName("diff")    DIFF("diff"),
        @SerializedName("browse")  BROWSE("browse"),
        @SerializedName("top")     TOP("top"),
        @SerializedName("verify")  VERIFY("verify"),
        @SerializedName("pin")     PIN("pin"),
        @SerializedName("clone")   CLONE("clone"),
        @SerializedName("rollback") ROLLBACK("rollback"),
        @SerializedName("cancel")  CANCEL("cancel"),
        @SerializedName("forecast") FORECAST("forecast"),
    }

    // --- Message Types ---
    data class Request(
        val tx_id: String = UUID.randomUUID().toString().take(8),
        val op: String,               // OpCode code
        val params: Map<String, Any?> = emptyMap()
    )

    data class Response(
        val tx_id: String,
        val status: String,           // "ok" | "error" | "progress"
        val message: String? = null,
        val data: JsonObject? = null
    )

    data class ProgressMessage(
        val tx_id: String,
        val phase: String,            // "snapshot", "scanning", "chunking", "compressing", "encrypting", "uploading", "committing"
        val progress: List<PhaseProgress>
    )

    data class PhaseProgress(
        val phase: String,
        val percent: Double,
        val files_done: Long,
        val files_total: Long,
        val bytes_done: Long,
        val bytes_total: Long
    )

    data class StatusData(
        val running: Boolean,
        val current_tx: String?,
        val state: String,            // "idle", "backing_up", "restoring", "verifying"
        val tps: Double,
        val cpu_percent: Double,
        val memory_mb: Long,
        val disk_iops_read: Long,
        val disk_iops_write: Long,
        val network_upload_mbps: Double,
        val queue_status: QueueStatus,
        val storage_stats: StorageStats?
    )

    data class QueueStatus(
        val scanner: Int,
        val chunk: Int,
        val compress: Int,
        val encrypt: Int,
        val upload: Int
    )

    data class StorageStats(
        val total_snapshots: Long,
        val total_size_bytes: Long,
        val dedup_ratio: Double,
        val packfile_count: Long
    )

    // --- Serialization helpers ---
    fun toJson(obj: Any): String = GSON.toJson(obj)

    fun parseRequest(json: String): Request = GSON.fromJson(json, Request::class.java)

    fun parseResponse(json: String): Response = GSON.fromJson(json, Response::class.java)

    fun parseProgress(json: String): ProgressMessage = GSON.fromJson(json, ProgressMessage::class.java)

    fun parseStatusData(data: JsonObject): StatusData = GSON.fromJson(data, StatusData::class.java)

    /**
     * Parameter builders for each operation.
     */
    object Params {
        fun backup(
            tag: String? = null,
            worldPath: String? = null,
            incremental: Boolean = true
        ): Map<String, Any?> = mapOf(
            "tag" to tag,
            "world_path" to worldPath,
            "incremental" to incremental,
            "timestamp" to System.currentTimeMillis()
        )

        fun status(): Map<String, Any?> = emptyMap<String, Any?>()

        fun restore(
            snapshotId: String,
            filePath: String? = null,
            chunkCoord: String? = null
        ): Map<String, Any?> = mapOf(
            "snapshot_id" to snapshotId,
            "file_path" to filePath,
            "chunk_coord" to chunkCoord
        )

        fun top(limit: Int = 5): Map<String, Any?> = mapOf("limit" to limit)

        fun diff(idA: String, idB: String): Map<String, Any?> = mapOf("id_a" to idA, "id_b" to idB)

        fun browse(snapshotId: String, path: String? = null): Map<String, Any?> = mapOf(
            "snapshot_id" to snapshotId,
            "path" to path
        )

        fun verify(repair: Boolean = false): Map<String, Any?> = mapOf("repair" to repair)

        fun pin(snapshotId: String, days: Int): Map<String, Any?> = mapOf(
            "snapshot_id" to snapshotId,
            "days" to days
        )

        fun clone(snapshotId: String, newName: String): Map<String, Any?> = mapOf(
            "snapshot_id" to snapshotId,
            "new_name" to newName
        )

        fun rollback(duration: String): Map<String, Any?> = mapOf("duration" to duration)

        fun forecast(): Map<String, Any?> = emptyMap<String, Any?>()
    }
}
